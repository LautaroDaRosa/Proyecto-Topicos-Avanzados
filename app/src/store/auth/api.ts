import { axiosInstance } from 'utils/axios';
import { toCamel, toSnake } from 'utils/convert-keys';
import TokenService from 'utils/tokenService';
import { LOGIN, REFRESH } from './endpoints';

import { LoginDataType } from './types';
import { Profile } from 'types';

const login = async (username, password) => {
  const response = await axiosInstance.post(LOGIN, {
    username,
    password,
  });
  return toCamel(response.data as LoginDataType);
};

const refresh = async () => {
  const data = {
    refresh: TokenService.getLocalRefreshToken(),
    grantType: 'refresh_token',
  };
  const response = await axiosInstance.post(REFRESH, toSnake(data));
  return toCamel(response.data);
};

export const getMyProfile = async () => {
  const url = `/user/my_profile`;
  const response = await axiosInstance.get(url);
  return toCamel(response.data as Profile);
};

const authApi = {
  login,
  refresh,
  getMyProfile,
};

export default authApi;
