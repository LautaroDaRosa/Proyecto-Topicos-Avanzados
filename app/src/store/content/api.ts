import { axiosInstance } from 'utils/axios';
import { toCamel } from 'utils/convert-keys';
import { Post } from './types';
import { PICTURES, POSTS } from './endpoints';
import { Picture } from 'types';

export const getPosts = async () => {
  const response = await axiosInstance.get(POSTS);
  return toCamel(response.data as Post[]);
};

export const addPost = async payload => {
  const response = await axiosInstance.post(POSTS, payload);
  return toCamel(response.data as Post);
};

export const getPictures = async () => {
  const response = await axiosInstance.get(PICTURES);
  return response.data as Picture[];
};

export const deletePicture = async id => {
  await axiosInstance.delete(`${PICTURES}${id}/`);
};

const authApi = {
  getPosts,
  addPost,
  getPictures,
  deletePicture,
};

export default authApi;
