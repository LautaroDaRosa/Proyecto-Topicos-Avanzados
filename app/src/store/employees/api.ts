import { axiosInstance } from 'utils/axios';
import { USERS } from './endpoints';
import { EmployeeWithPosts, MinimalEmployee } from 'types';
import { toCamel } from 'utils/convert-keys';

export const getEmployee = async id => {
  const response = await axiosInstance.get(`${USERS + id}/`);
  return toCamel(response.data as EmployeeWithPosts);
};

export const getEmployees = async () => {
  const response = await axiosInstance.get(USERS);
  return response.data as MinimalEmployee[];
};

export const getPermissions = async () => {
  const response = await axiosInstance.get(`${USERS}permissions/`);
  return response.data as string[];
};

const authApi = {
  getEmployee,
  getEmployees,
  getPermissions,
};

export default authApi;
