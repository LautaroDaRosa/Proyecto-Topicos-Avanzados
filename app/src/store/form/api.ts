import { Question, QuestionWithId } from 'types';
import { axiosInstance } from 'utils/axios';
import { toCamel } from 'utils/convert-keys';
import { ANSWER, QUESTION } from './endpoints';
import { QuestionAnswer } from './types';

export const postQuestions = async (questions: Question[]) => {
  const response = await axiosInstance.post(QUESTION, questions);
  return toCamel(response.data);
};

export const getQuestions = async () => {
  const response = await axiosInstance.get(QUESTION);
  return toCamel(response.data as QuestionWithId[]);
};

export const sendAnswers = async (answers: QuestionAnswer[]) => {
  const response = await axiosInstance.post(`${ANSWER}/send`, answers);
  return toCamel(response.data);
};

const authApi = {
  postQuestions,
  getQuestions,
  sendAnswers,
};

export default authApi;
