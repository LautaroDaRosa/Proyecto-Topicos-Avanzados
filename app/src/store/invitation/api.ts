import { axiosInstance } from 'utils/axios';
import { toCamel } from 'utils/convert-keys';
import { INVITE } from './endpoints';
import { Invitation } from './types';

export const sendInvitation = async (email: string) => {
  const response = await axiosInstance.post(`${INVITE}/send?email=${email}`);
  return toCamel(response.data);
};

export const getInvitation = async () => {
  const response = await axiosInstance.get(`${INVITE}/get`);
  return response.data as Invitation[];
};

export const respondInvitation = async (status: 'ACCEPTED' | 'REJECTED') => {
  await axiosInstance.post(`${INVITE}/respond?status=${status}`);
};

const authApi = {
  sendInvitation,
  getInvitation,
};

export default authApi;
