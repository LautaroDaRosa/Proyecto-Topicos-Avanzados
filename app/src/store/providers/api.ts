import { axiosInstance } from 'utils/axios';
import { toCamel } from 'utils/convert-keys';
import { Provider } from './types';
import { CATEGORIES, PROVIDERS } from './endpoints';

export const getProviders = async () => {
  const response = await axiosInstance.get(PROVIDERS);
  return toCamel(response.data as Provider[]);
};

export const getCategories = async () => {
  const response = await axiosInstance.get(CATEGORIES);
  return response.data as string[];
};

const authApi = {
  getProviders,
  getCategories,
};

export default authApi;
