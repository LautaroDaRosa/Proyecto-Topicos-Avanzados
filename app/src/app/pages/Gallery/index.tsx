import Navbar from 'app/components/NavBar';
import { Picture } from 'types';
import StGridsContainer from '../../components/PhotoGrid/StGridsContainer';
import StFlex from 'app/components/Flex';
import { deletePicture, getPictures } from 'store/content/api';
import { useEffect, useState } from 'react';
import { getPermissions } from 'store/employees/api';
import PhotoGrid from 'app/components/PhotoGrid';

const Gallery = () => {
  const [pictures, setPictures] = useState<Picture[]>([]);
  const featuredPics: Picture[] = [];
  const notFeaturedPics: Picture[] = [];
  const [hasPerm, setHasPerm] = useState(false);

  useEffect(() => {
    async function fetchPictures() {
      const result = await getPictures();
      setPictures(result);
    }
    fetchPictures();
  }, []);

  useEffect(() => {
    async function getPermission() {
      const result = await getPermissions();
      setHasPerm(result.includes('content.delete_picture'));
    }
    getPermission();
  }, []);

  pictures.forEach(pict =>
    pict.featured ? featuredPics.push(pict) : notFeaturedPics.push(pict),
  );

  if (featuredPics.length % 2 !== 0) {
    const picture = notFeaturedPics.shift();
    if (picture !== undefined) {
      featuredPics.push(picture);
    }
  }

  const deletePict = async (id: number) => {
    await deletePicture(id);
    setPictures(pictures.filter(pic => pic.id !== id));
  };

  return (
    <StFlex flexProps={{}}>
      <Navbar />
      <StGridsContainer>
        <PhotoGrid
          isFeatured={true}
          pics={featuredPics}
          hasPerm={hasPerm}
          deletePict={deletePict}
        />
        <PhotoGrid
          isFeatured={false}
          pics={notFeaturedPics}
          hasPerm={hasPerm}
          deletePict={deletePict}
        />
      </StGridsContainer>
    </StFlex>
  );
};

export default Gallery;
