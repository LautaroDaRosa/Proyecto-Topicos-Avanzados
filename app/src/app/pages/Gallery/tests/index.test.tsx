import { fireEvent, screen, waitFor } from '@testing-library/react';
import api from '../../../../store/content/api';
import employeeApi from '../../../../store/employees/api';
import Gallery from '..';
import renderWithWrappers from 'utils/render';

jest.mock('../../../../store/content/api');
jest.mock('../../../../store/employees/api');

describe('Gallery', () => {
  it('shows pictures fetched from the backend correctly', async () => {
    (api.getPictures as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          id: 1,
          title: "A day at Octobot's office",
          photo: 'pictures/pic1.jpeg',
          featured: true,
        },
        {
          id: 2,
          title: 'Octobot Office',
          photo: 'pictures/pic2.jpeg',
          featured: true,
        },
        {
          id: 3,
          title: 'Octobot Team',
          photo: 'pictures/pic3.jpeg',
          featured: false,
        },
      ]),
    );

    (employeeApi.getPermissions as jest.Mock).mockReturnValue(
      Promise.resolve([]),
    );

    const { container } = renderWithWrappers(<Gallery />);

    expect(api.getPictures).toHaveBeenCalledWith();

    await screen.findByText('Octobot Office');
    screen.getByText("A day at Octobot's office");

    const img1 = screen.getByRole('img', { name: "A day at Octobot's office" });
    const img2 = screen.getByRole('img', { name: 'Octobot Office' });
    const img3 = screen.getByRole('img', { name: 'Octobot Team' });

    expect(img1.getAttribute('src')).toBe('pictures/pic1.jpeg');
    expect(img2.getAttribute('src')).toBe('pictures/pic2.jpeg');
    expect(img3.getAttribute('src')).toBe('pictures/pic3.jpeg');

    expect(container).toMatchSnapshot();
  });

  it('displays the delete button when user is allowed', async () => {
    (api.getPictures as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          id: 1,
          title: "A day at Octobot's office",
          photo: 'pictures/pic1.jpeg',
          featured: true,
        },
      ]),
    );

    (employeeApi.getPermissions as jest.Mock).mockReturnValue(
      Promise.resolve(['content.delete_picture']),
    );

    renderWithWrappers(<Gallery />);

    expect(employeeApi.getPermissions).toHaveBeenCalledWith();
    await screen.findByRole('img', { name: "A day at Octobot's office" });

    screen.getByRole('button');
  });

  it('does not display the delete button when user is not allowed', async () => {
    (api.getPictures as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          id: 1,
          title: "A day at Octobot's office",
          photo: 'pictures/pic1.jpeg',
          featured: true,
        },
      ]),
    );

    (api.deletePicture as jest.Mock).mockReturnValue(Promise.resolve({}));

    (employeeApi.getPermissions as jest.Mock).mockReturnValue(
      Promise.resolve([]),
    );

    renderWithWrappers(<Gallery />);

    expect(employeeApi.getPermissions).toHaveBeenCalledWith();
    await screen.findByRole('img', { name: "A day at Octobot's office" });

    const btn = screen.queryByRole('button');
    expect(btn).toBeNull();
  });

  it('deletes the picture correctly', async () => {
    (api.getPictures as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          id: 1,
          title: "A day at Octobot's office",
          photo: 'pictures/pic1.jpeg',
          featured: true,
        },
      ]),
    );

    (api.deletePicture as jest.Mock).mockReturnValue(Promise.resolve({}));

    (employeeApi.getPermissions as jest.Mock).mockReturnValue(
      Promise.resolve(['content.delete_picture']),
    );

    renderWithWrappers(<Gallery />);

    expect(employeeApi.getPermissions).toHaveBeenCalledWith();

    await screen.findByRole('img', { name: "A day at Octobot's office" });

    const btn = screen.getByRole('button', { name: 'Delete' });

    fireEvent.click(btn);

    await waitFor(() =>
      expect(
        screen.queryByRole('img', {
          name: "A day at Octobot's office",
        }),
      ).toBeNull(),
    );
  });
});
