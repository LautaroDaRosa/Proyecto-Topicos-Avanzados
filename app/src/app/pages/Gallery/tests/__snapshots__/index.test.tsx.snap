// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery shows pictures fetched from the backend correctly 1`] = `
.c1 {
  height: 72px;
  background-color: white;
  box-shadow: 0px 0px 20px 0px lightgrey;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  width: 100%;
  left: 0;
  top: 0;
  z-index: 9;
}

.c1 img {
  height: 100%;
}

.c4 {
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  padding-left: 40px;
}

.c4 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #1c1a46;
  font-size: 16px;
  font-style: normal;
  font-weight: 600;
  line-height: normal;
  text-transform: capitalize;
}

.c3 {
  background: #c0bfcb;
  width: 1px;
  height: 24px;
  border-radius: 1px;
  display: none;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 12px 24px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 8px;
  margin-right: 20px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  justify-items: center;
  cursor: default;
}

.c5 span {
  font-style: normal;
  font-weight: 400;
  line-height: normal;
  text-transform: capitalize;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 14px;
}

.c5 img {
  width: 24px;
}

.c5 span {
  color: #c70548;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 12px 24px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 8px;
  margin-right: 20px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  justify-items: center;
}

.c6 span {
  font-style: normal;
  font-weight: 400;
  line-height: normal;
  text-transform: capitalize;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 14px;
}

.c6 img {
  width: 24px;
}

.c6 span {
  color: #1c1a46;
}

.c6:hover {
  background-color: #f1f1f5;
}

.c2 {
  padding-right: 50px;
  font-size: 20px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  gap: 10px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 88px 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: px;
  width: %;
}

.c8 {
  display: grid;
  width: 100%;
  gap: 10px;
  grid-template-columns: 1fr;
}

.c11 {
  display: grid;
  width: 100%;
  gap: 10px;
  grid-template-columns: 1fr;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: hidden;
}

.c9:hover span {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 img {
  width: 100%;
  -webkit-transition: filter 0.1s;
  transition: filter 0.1s;
  object-fit: cover;
  -webkit-transition: -webkit-transform 0.3s ease;
  -webkit-transition: transform 0.3s ease;
  transition: transform 0.3s ease;
}

.c9:hover > img {
  -webkit-filter: brightness(75%);
  filter: brightness(75%);
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c9:hover div {
  opacity: 100%;
}

.c10 {
  position: absolute;
  color: white;
  bottom: 16px;
  left: 12px;
  font-size: 16px;
  display: none;
  font-weight: 300;
}

@media (max-width:1279px) {
  .c4 {
    position: fixed;
    bottom: 0px;
    width: 100%;
    background: white;
    height: 72px;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (min-width:1280px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media (min-width:1280px) {
  .c5 img {
    display: none;
  }

  .c5 span {
    font-size: 20px;
  }
}

@media (min-width:1280px) {
  .c6 img {
    display: none;
  }

  .c6 span {
    font-size: 20px;
  }
}

@media (min-width:768px) {
  .c7 {
    padding: 94px 24px;
  }
}

@media (min-width:1280px) {
  .c7 {
    padding: 156px 156px 84px 156px;
  }
}

@media (min-width:768px) {
  .c8 {
    grid-template-columns: 1fr 1fr;
  }
}

@media (min-width:768px) {
  .c11 {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

<div>
  <div
    class="c0"
  >
    <header
      class="c1"
    >
      <img
        alt="logo"
        src="logo.png"
      />
      <span
        class="c2"
      >
        Octobot Community
      </span>
      <div
        class="c3"
      />
      <nav
        class="c4"
      >
        <a
          href="/"
        >
          <div
            class="c5"
          >
            <img
              alt="Home icon"
              src="home-selected.svg"
            />
            <span>
              Home
            </span>
          </div>
        </a>
        <a
          href="/gallery"
        >
          <div
            class="c6"
          >
            <img
              alt="Gallery icon"
              src="gallery.svg"
            />
            <span>
              Gallery
            </span>
          </div>
        </a>
      </nav>
    </header>
    <div
      class="c7"
    >
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <img
            alt="A day at Octobot's office"
            src="pictures/pic1.jpeg"
          />
          <span
            class="c10"
          >
            A day at Octobot's office
          </span>
        </div>
        <div
          class="c9"
        >
          <img
            alt="Octobot Office"
            src="pictures/pic2.jpeg"
          />
          <span
            class="c10"
          >
            Octobot Office
          </span>
        </div>
      </div>
      <div
        class="c11"
      >
        <div
          class="c9"
        >
          <img
            alt="Octobot Team"
            src="pictures/pic3.jpeg"
          />
          <span
            class="c10"
          >
            Octobot Team
          </span>
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
